clients/src/main/java/org/apache/kafka/common/network/SslTransportLayer.java

```diff
@@ -429,27 +493,30 @@ private SSLEngineResult handshakeUnwrap(boolean doRead) throws IOException {
     *
     * @param dst The buffer into which bytes are to be transferred
     * @return The number of bytes read, possible zero or -1 if the channel has reached end-of-stream
+    *         and no more data is available
     * @throws IOException if some other I/O error occurs
     */
     @Override
     public int read(ByteBuffer dst) throws IOException {
-        if (closing) return -1;
         int read = 0;
-        if (!handshakeComplete) return read;
+        if (state == State.CLOSING) return -1;
+        else if (state != State.READY) return read;
 
         //if we have unread decrypted data in appReadBuffer read that into dst buffer.
         if (appReadBuffer.position() > 0) {
             read = readFromAppBuffer(dst);
         }
 
+        int netread = 0;
         if (dst.remaining() > 0) {
             netReadBuffer = Utils.ensureCapacity(netReadBuffer, netReadBufferSize());
             if (netReadBuffer.remaining() > 0) {
```

Comment:
Can this check ever fail?

Reply:
Don't think so, but I didn't want to remove it just in case.